lib/FrontendTool/FrontendTool.cpp

```diff
@@ -184,73 +185,144 @@ static bool emitMakeDependenciesIfNeeded(DiagnosticEngine &diags,
 }
 
 namespace {
+struct SwiftModuleTraceInfo {
+  Identifier Name;
+  std::string Path;
+  bool IsImportedDirectly;
+  bool SupportsLibraryEvolution;
+};
+
 struct LoadedModuleTraceFormat {
-  std::string Name;
+  static const unsigned CurrentVersion = 2;
+  unsigned Version;
+  Identifier Name;
   std::string Arch;
-  std::vector<std::string> SwiftModules;
+  std::vector<SwiftModuleTraceInfo> SwiftModules;
 };
 }
 
 namespace swift {
 namespace json {
+template <> struct ObjectTraits<SwiftModuleTraceInfo> {
+  static void mapping(Output &out, SwiftModuleTraceInfo &contents) {
+    std::string name(contents.Name.get(), contents.Name.getLength());
+    out.mapRequired("name", name);
```

Comment:
I believe LLVM's YAML support allows for StringRefs in `mapRequired`, so we can avoid a copy here. (Not that it *really* matters, but you might as well.)

Reply:
I tried that earlier and StringRef doesn't work. My understanding (based on discussion with Harlan) is that you need to pass a reference to a type which has ownership (as parsing will modify the address), so it doesn't make sense for it to work with StringRef.

Reply:
You can't pass the StringRef directly, but you *can* pass a reference to a local variable StringRef. (All of this only works because we don't try to use the mapping to read values in, only to write them out. Otherwise, your temporary strings wouldn't work either.)

Reply:
So this is one of those cases where let-expansion makes code compile when it wouldn't compile otherwise. Yikes, fixed. 🤕

Reply:
If we *were* trying to use this to parse too, it'd be important!

Reply:
You got one of them but not the other one…